app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperDownloader.kt

```diff
@@ -36,41 +37,59 @@ class WallpaperDownloader(
      * and will be stored in the local path:
      * wallpapers/<wallpaper name>/<orientation>.png
      */
-    suspend fun downloadWallpaper(wallpaper: Wallpaper) = withContext(dispatcher) {
-        for (metadata in wallpaper.toMetadata()) {
-            val localFile = File(storageRootDirectory.absolutePath, metadata.localPath)
-            // Don't overwrite an asset if it exists
-            if (localFile.exists()) continue
-            val request = Request(
-                url = "$remoteHost/${metadata.remotePath}",
-                method = Request.Method.GET
-            )
-            Result.runCatching {
-                val response = client.fetch(request)
-                if (!response.isSuccess) {
-                    return@withContext
-                }
-                File(localFile.path.substringBeforeLast("/")).mkdirs()
-                response.body.useStream { input ->
-                    input.copyTo(localFile.outputStream())
-                }
-            }.onFailure {
-                // This should clean up any partial downloads
-                Result.runCatching {
-                    if (localFile.exists()) {
-                        localFile.delete()
-                    }
-                }
-            }
+    suspend fun downloadWallpaper(wallpaper: Wallpaper): Wallpaper.ImageFileState = withContext(dispatcher) {
+        val portraitResult = wallpaper.downloadAsset(Wallpaper.ImageType.Portrait)
+        val landscapeResult = wallpaper.downloadAsset(Wallpaper.ImageType.Landscape)
+        return@withContext if (portraitResult == Wallpaper.ImageFileState.Downloaded &&
+            landscapeResult == Wallpaper.ImageFileState.Downloaded
+        ) {
+            Wallpaper.ImageFileState.Downloaded
+        } else {
+            Wallpaper.ImageFileState.Error
         }
     }
 
-    private data class WallpaperMetadata(val remotePath: String, val localPath: String)
+    /**
+     * Downloads a thumbnail for a wallpaper from the network. This is expected to be found remotely
+     * at:
+     * <WALLPAPER_URL>/<collection name>/<wallpaper name>/thumbnail.png
+     * and stored locally at:
+     * wallpapers/<wallpaper name>/thumbnail.png
+     */
+    suspend fun downloadThumbnail(wallpaper: Wallpaper): Wallpaper.ImageFileState = withContext(dispatcher) {
+        wallpaper.downloadAsset(Wallpaper.ImageType.Thumbnail)
+    }
+
+    private suspend fun Wallpaper.downloadAsset(
```

Comment:
I updated this to be more clear, thanks for the suggestion. I think it would be worth bringing up with the team whether we want to have any specific style guides for when to use extensions or not